File: 42-trapping-rain-water/trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    let left = new Array(height.length)
    let right = new Array(height.length)
    left[0] = height[0]
    right[height.length-1] = height[height.length-1]
    let maxleft = height[0]
    let maxright = height[height.length-1]
    for(let i = 1; i<height.length; i++){
        maxleft = Math.max(maxleft, height[i])
        left[i] = maxleft
    }
    for(let i = height.length-2; i>=0; i--){
        maxright = Math.max(maxright, height[i])
        right[i] = maxright
    }
    let ans = 0
    for(let i = 0; i< height.length; i++){
        ans += Math.min(left[i], right[i]) - height[i]
    }
    return ans

};

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        left = [0]* len(height)
        right = [0]* len(height)
        left[0] = height[0]
        right[len(height)-1] = height[len(height)-1]
        maxleft = height[0]
        maxright = height[len(height)-1]
        for i in range(1, len(left)):
            if(height[i] > maxleft): 
                maxleft = height[i]
                left[i] = maxleft
            else: left[i] = maxleft
        

        for i in range(len(right)-2, -1, -1):
            if(height[i]> maxright): 
                maxright = height[i]
                right[i] = maxright
            else: right[i] = maxright
        
        ans = 0
        for i in range(0, len(height)):
            ans += min(left[i], right[i]) - height[i]
        return ans
        